<a href="https://colab.research.google.com/github/Andreagazy/Machine-Learning-2024/blob/main/Praktikum%204%20-%20Klasifikasi/Tugas_Klasifikasi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tugas 1

## 1. Buatlah model klasifikasi dengan menggunakan SVM untuk data suara, `voice.csv`.

## Import library

In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import classification_report, accuracy_score
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Data Load



In [5]:
data = pd.read_csv('/content/drive/MyDrive/Machine Learning/voice.csv')

# Preprocessing

In [11]:
X = data.drop(columns='label', axis = 1)  # Mengasumsikan kolom 'label' adalah target
y = data['label']

## Splitting data

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


# Model

In [16]:
svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)

SVC(kernel='linear')

# Evaluasi

In [19]:
y_pred = svm_model.predict(X_test)

# Print evaluation metrics
print(f"Akurasi : {accuracy_score(y_test, y_pred) * 100:.2f}%")
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy: 97.63%
Classification Report:
               precision    recall  f1-score   support

      female       0.96      0.99      0.98       297
        male       0.99      0.97      0.98       337

    accuracy                           0.98       634
   macro avg       0.98      0.98      0.98       634
weighted avg       0.98      0.98      0.98       634



## 2. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

  1. Menggunakan data `spam.csv`
  2. Fitur `CountVectorizer` dengan mengaktifkan `stop_words`
  3. Evaluasi hasilnya

## Data Load

In [45]:
data = pd.read_csv('/content/drive/MyDrive/Machine Learning/spam.csv',  encoding='latin-1')

data = data[['v1', 'v2']]
data.columns = ['label', 'text']
X = data['text']
y = data['label']

## Splitting data & Stop Words

In [46]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Buat dan latih model Multinomial Naive Bayes
vectorizer = CountVectorizer(stop_words='english')


X_train_cv = vectorizer.fit_transform(X_train)
X_test_cv = vectorizer.transform(X_test)


## Model

In [47]:
nb_model = MultinomialNB()
nb_model.fit(X_train_cv, y_train)

MultinomialNB()

## Evaluasi

In [48]:
y_pred_cv = nb_model.predict(X_test_cv)

print(f"Accuracy (Count Vectorizer): {accuracy_score(y_test, y_pred_cv) * 100:.2f}%")
print("Classification Report:\n", classification_report(y_test, y_pred_cv))

Accuracy (Count Vectorizer): 98.39%
Classification Report:
               precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



## 3. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

  1. Menggunakan data `spam.csv`

  2. Fitur `TF-IDF` dengan mengaktifkan `stop_words`

  3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

  4. Berikan kesimpulan fitur mana yang terbaik pada kasus data `spam.csv`

## Data Load

In [37]:
data = pd.read_csv('/content/drive/MyDrive/Machine Learning/spam.csv',  encoding='latin-1')

data = data[['v1', 'v2']]
data.columns = ['label', 'text']
X = data['text']
y = data['label']

## Splitting data & Stop Words

In [38]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

tfidf_vectorizer = TfidfVectorizer(stop_words='english')
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)



## Model

In [39]:
nb_model_tfidf = MultinomialNB()
nb_model_tfidf.fit(X_train_tfidf, y_train)

MultinomialNB()

## Evaluasi

In [30]:
y_pred_tfidf = nb_model_tfidf.predict(X_test_tfidf)

# Menampilkan akurasi dan classification report
print(f"Accuracy (TF-IDF): {accuracy_score(y_test, y_pred_tfidf) * 100:.2f}%")
print("Classification Report (TF-IDF):\n", classification_report(y_test, y_pred_tfidf))

Accuracy (TF-IDF): 96.68%
Classification Report (TF-IDF):
               precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



## Perbandingan dengan no 2

In [44]:
print(f"Accuracy (Count Vectorizer): {accuracy_score(y_test, y_pred_cv) * 100:.2f}%")
print("Classification Report:\n", classification_report(y_test, y_pred_cv))
print()
print(f"Accuracy (TF-IDF): {accuracy_score(y_test, y_pred_tfidf) * 100:.2f}%")
print("Classification Report (TF-IDF):\n", classification_report(y_test, y_pred_tfidf))

Accuracy (Count Vectorizer): 98.39%
Classification Report:
               precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115


Accuracy (TF-IDF): 96.68%
Classification Report (TF-IDF):
               precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



## Kesimpulan Hasil
Model `TF-IDF` menunjukkan `precision` yang sangat baik untuk kelas spam (1.00), tetapi memiliki `precision` yang lebih rendah untuk kelas ham (0.96). Di sisi lain, `Count Vectorizer` memiliki `precision` yang lebih seimbang antara kedua kelas. <br><br>
`Count Vectorizer` memiliki `recall` yang lebih baik untuk spam (0.92) dibandingkan `TF-IDF` (0.75) untuk kelas spam. Namun, `TF-IDF` memiliki `recall` yang sempurna untuk kelas ham (1.00). <br><br>
`F1-Score` untuk spam lebih tinggi pada `Count Vectorizer` (0.94) dibandingkan `TF-IDF` (0.86). Ini menunjukkan bahwa `Count Vectorizer` lebih baik dalam menangani trade-off antara `precision` dan `recall`. <br><br>
Model dengan `Count Vectorizer` memiliki akurasi yang lebih tinggi (98.39%) dibandingkan model dengan `TF-IDF` (96.68%). Ini menunjukkan bahwa Count Vectorizer lebih efektif dalam menangani data ini.



# Tugas 2

## Buatlah model klasifikasi dengan menggunakan SVM berdasarkan data pada Percobaan 5 dengan menggunakan fitur histogram.